v18 backport: `schemadiff`: fix diffing of textual columns with implicit charsets

shlomi-noach opened the pull request:
## Description

A manual (and partial) backport of https://github.com/vitessio/vitess/pull/14930. This had to be made manually because the changes in `v19` rely on un-backportable changes we made to the collation environment.

This fixes an issue in `v18`, described in https://github.com/vitessio/vitess/pull/15859#issuecomment-2099325284

## Related Issue(s)

https://github.com/vitessio/vitess/pull/15859#issuecomment-2099325284

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

## Deployment Notes

<!-- Notes regarding deployment of the contained body of work. These should note any db migrations, etc. -->

arthurschreiber commented:
> Seems like `Cluster (schemadiff_vrepl) mysql57` CI test fails consistently. Looking.

I saw this on my branch too. I think MySQL 5.7 and MySQL 8 had different behaviour around when and how charset and collation were part of the table definition.

arthurschreiber commented:
If I remember correctly, this might be related to the default charset on MySQL 5.7 being different from 8.0.

So, on 5.7, the table will be created with `utf8` as the default charset, and `utf8` as the charset on the columns. Because the charset on the columns and the charset on the table matches, MySQL will skip including the charset in the `SHOW CREATE TABLE` output for the columns, and also skip listing it on the table itself (because it matches the system charset).

shlomi-noach commented:
That makes sense. I haven't had the time to dig into this deeply yet (about to work on the point release now) but will take a look within an hour or so.

shlomi-noach commented:
Updated `vitess-private` to include https://github.com/planetscale/vitess-private/pull/5131, and now using `AtomicDiffs()` function to split the `DROP PARTITION` statements. Removed a bunch of code.

arthurschreiber commented:
Seems like this is related to the alias?

Here's another failing test case: ```
		{
			name: "two identical tables, one with canonical charset name, one with alias",
			fromQueries: []string{
				"create table foobar (id int primary key, foo varchar(64) character set utf8 collate utf8_bin)",
			},
			toQueries: []string{
				"create table foobar (id int primary key, foo varchar(64) character set utf8mb3 collate utf8mb3_bin)",
			},
		},
```

shlomi-noach commented:
A bit struggling with the test failure and how to address it.

shlomi-noach commented:
> Here's another failing test case:

Thanks for the pointer. FWIW, the best place for this kind of test is in `TestCreateTableDiff`, in `table_test.go`.

shlomi-noach commented:
> Here's another failing test case:

This is passing for me in this PR though.

shlomi-noach commented:
OK, there's yet more code to backport, in particular `getTableCharsetCollate()`

shlomi-noach commented:
I feel like this is the very problem we only fixed in v19 and could not backport to v18. The multitude of MySQL versions and their contextual environment.

shlomi-noach commented:
If I add the fix, that break 8.0 behavior.

arthurschreiber commented:
Maybe I can work around this on my side a bit by also setting the charset on the table itself?

shlomi-noach commented:
Seems like we won't be needing this PR after all, since https://github.com/vitessio/vitess/pull/15859 found a simpler solution. Closing for now.
